lib/virtual-fs/src/mem_fs.rs

```diff
@@ -0,0 +1,600 @@
+use crate::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::io::{Read, Seek, Write};
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+
+pub use crate::host_fs::{Stderr, Stdin, Stdout};
+
+#[derive(Debug, Clone)]
+enum MemKind {
+    File {
+        name: String,
+        inode: u64,
+    },
+    Directory {
+        name: String,
+        contents: HashMap<String, MemKind>,
+    },
+}
+
+impl Default for MemKind {
+    fn default() -> Self {
+        MemKind::Directory {
+            name: "/".to_string(),
+            contents: Default::default(),
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct MemFileSystem {
+    inner: Arc<Mutex<MemFileSystemInner>>,
+}
+
+#[derive(Default)]
+pub struct MemFileSystemInner {
+    // done for recursion purposes
+    fs: MemKind,
+    inodes: HashMap<u64, Box<dyn VirtualFile>>,
```

Comment:
I think inodes should not be needed for a MemoryFilesystem? Or is this designed with symlinks in mind?

Why do we need the MemFileHandle?

Reply:
It seems we have `MemFileHandle` and `MemFile` as two ways of having a file in the memfs. I'm still trying to wrap my head around on why do we need two?

Reply:
It's a bit of a band-aid to cover up a mismatch: the API is written in terms of `Box<dyn VirtualFile>` but because it's also the backing storage we need to share it. So the internal file implementation here (which is also implemented in terms of `VirtualFile` but doesn't have to be) is shared with `Arc` and we return a `Box<dyn VirtualFile>` for the API itself. This can be fixed after more iteration, I just didn't want to make too many big changes all at once